File: Data Structures/Trees/Depth-First Search/230/230.go
package main

// Definition for a binary tree node.
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// kthSmallest finds the kth smallest element in a binary search tree.
// The function takes in a pointer to the root of the binary search tree
// and an integer k indicating the kth smallest element to find.
// It returns an integer representing the kth smallest element.
func kthSmallest(root *TreeNode, k int) int {
	// Create a stack to store the nodes of the binary search tree.
	// Initialize the stack with a capacity of k.
	stack := make([]*TreeNode, 0, k)

	// Traverse the binary search tree in an in-order manner.
	// When we encounter a node, push it onto the stack.
	// Keep traversing the left subtree until we reach a leaf node.
	for {
		// Traverse the left subtree until we reach a leaf node.
		for root != nil {
			// Push the current node onto the stack.
			stack = append(stack, root)
			// Move to the left subtree.
			root = root.Left
		}

		// If the stack is empty, it means we have reached the end of
		// the binary search tree.
		if len(stack) == 0 {
			break
		}

		// Pop the top node from the stack.
		root = stack[len(stack)-1]
		stack = stack[:len(stack)-1]

		// Decrement the value of k.
		k--

		// If k becomes 0, it means we have found the kth smallest element.
		if k == 0 {
			return root.Val
		}

		// Traverse the right subtree.
		root = root.Right
	}

	// If we reach this point, it means the binary search tree is empty or
	// the value of k is greater than the number of nodes in the tree.
	return -1
}

// Time Complexity: O(n)
// Space Complexity: O(n)

File: Data Structures/Trees/Depth-First Search/230/230.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        """
        Initialize a new TreeNode with the given value and left and right child nodes.
        
        Args:
            val (int): The value to be assigned to the TreeNode.
            left (TreeNode): The left child of the TreeNode (default: None).
            right (TreeNode): The right child of the TreeNode (default: None).
        """
        # Assign the given value to the TreeNode's val attribute
        self.val = val
        
        # Assign the given left child to the TreeNode's left attribute
        self.left = left
        
        # Assign the given right child to the TreeNode's right attribute
        self.right = right

class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        # Initialize an empty stack to keep track of the nodes to be visited
        stack = []
        
        # Initialize a variable to keep track of the current node
        curr = root
        
        # Keep traversing the binary search tree until we have found the kth smallest element
        while stack or curr:
            
            # Traverse as far left as possible from the current node
            # This ensures that we are always considering the leftmost nodes first
            while curr:
                # Add the current node to the stack to be visited later
                stack.append(curr)
                
                # Move to the left child of the current node
                curr = curr.left
            
            # Pop the leftmost node from the stack and move to its right child
            # This ensures that we are always considering the right child next
            curr = stack.pop()
            
            # Decrement the count of nodes to be visited
            k -= 1
            
            # If we have found the kth smallest element, return its value
            if k == 0:
                return curr.val
            
            # Move to the right child of the current node
            curr = curr.right

# Time Complexity: O(N)
# Space Complexity: O(N)